server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -0,0 +1,153 @@
+        public static final String GET_SEGMENT_FILES = "internal:index/shard/replication/get_segment_files";
+    }
+
+    private final Map<ReplicationCheckpoint, CopyState> copyStateMap;
+    private final TransportService transportService;
+    private final IndicesService indicesService;
+
+    // TODO mark this as injected and bind in Node
+    public SegmentReplicationSourceService(TransportService transportService, IndicesService indicesService) {
+        copyStateMap = Collections.synchronizedMap(new HashMap<>());
+        this.transportService = transportService;
+        this.indicesService = indicesService;
+
+        transportService.registerRequestHandler(
+            Actions.GET_CHECKPOINT_INFO,
+            ThreadPool.Names.GENERIC,
+            CheckpointInfoRequest::new,
+            new CheckpointInfoRequestHandler()
+        );
+        transportService.registerRequestHandler(
+            Actions.GET_SEGMENT_FILES,
+            ThreadPool.Names.GENERIC,
+            GetSegmentFilesRequest::new,
+            new GetSegmentFilesRequestHandler()
+        );
+    }
+
+    private class CheckpointInfoRequestHandler implements TransportRequestHandler<CheckpointInfoRequest> {
+        @Override
+        public void messageReceived(CheckpointInfoRequest request, TransportChannel channel, Task task) throws Exception {
+            final ReplicationCheckpoint checkpoint = request.getCheckpoint();
+            logger.trace("Received request for checkpoint {}", checkpoint);
+            final CopyState copyState = getCachedCopyState(checkpoint);
+            channel.sendResponse(
+                new CheckpointInfoResponse(
+                    copyState.getCheckpoint(),
+                    copyState.getMetadataSnapshot(),
+                    copyState.getInfosBytes(),
+                    copyState.getPendingDeleteFiles()
+                )
+            );
+        }
+    }
+
+    class GetSegmentFilesRequestHandler implements TransportRequestHandler<GetSegmentFilesRequest> {
+        @Override
+        public void messageReceived(GetSegmentFilesRequest request, TransportChannel channel, Task task) throws Exception {
+            if (isInCopyStateMap(request.getCheckpoint())) {
+                // TODO send files
+            } else {
+                channel.sendResponse(TransportResponse.Empty.INSTANCE);
+            }
+        }
+    }
+
+    /**
+     * Operations on the {@link #copyStateMap} member.
+     */
+
+    /**
+     * A synchronized method that checks {@link #copyStateMap} for the given {@link ReplicationCheckpoint} key
+     * and returns the cached value if one is present. If the key is not present, a {@link CopyState}
+     * object is constructed and stored in the map before being returned.
+     */
+    private synchronized CopyState getCachedCopyState(ReplicationCheckpoint checkpoint) throws IOException {
```

Comment:
We'll need to handle replicas that shut down between these two requests.  Either by reacting to cluster state changes indicating a replica has left or a TTL.

Reply:
I'll cut a separate issue to track this